### DISCUSSION TITLE
Can ebiten make game mod?

### DISCUSSION THREAD

**Discussion by lvyonghuan** [hajimehoshi/ebiten#3038] [General]
I think using Go to create a game mod might be a bit challenging. Our small studio wants to enable the community to create functional extension mods for our game.

What I'm currently considering is using dynamic linking (since I’m working on Linux, using the plugin package could be a viable option) and employing middleware, exposing it as an API. Additionally, using a config file could be a good choice.

I wonder if there are any examples of creating mods for games made with a Go engine. Can someone share?

**Comment by SolarLune** [comment]
Late to reply, but my approach for my game (that I'm still working on) is to use goja for scripting, which should allow for modding the logic, and to hotload assets from the game directory, which should allow modding the visual and audio assets.

**Reply by lvyonghuan** [reply]
> Late to reply, but my approach for my game (that I'm still working on) is to use goja for scripting, which should allow for modding the logic, and to hotload assets from the game directory, which should allow modding the visual and audio assets.

Sensitive your answer, although we have temporarily left this matter behind.
I feel that for Go language, making mods is indeed not an easy task. It really needs other languages ​​to assist. Go's scenario is more about cloud services. It would be much more convenient if there was a dynamic linking mechanism, but Go official felt that this was not the main scenario and it was not continued to advance.

**Reply by SolarLune** [reply]
Well, dynamic loading of libraries would be nice, but still would only be a smaller part of the puzzle for games. Because Go is a compiled language, a potential mod creator would probably have to build libraries for all supported operating systems, which is simple in Go, but still not as simple as in a dynamic language like Lua or Python. The mod would, more critically, have access to all of the functionality a normal Go program has access to, like full access to your host filesystem and making unmonitored web calls. 

A really game modding-friendly programming language would probably need more security and locking features than most languages would ordinarily support.
